File: src/tokenizer.js
var assert = require('assert');
var {SyntaxError} = require('./error');

let escapeRegex = (s) => s.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');

let isRegEx = (re) => {
  if (typeof re !== 'string') {
    return false;
  }
  try {
    new RegExp(`^${re}$`);
  } catch (e) {
    return false;
  }
  return true;
};

let isNonCaptureRegex = (re) => {
  return isRegEx(re) && (new RegExp(`^(?:|${re})$`)).exec('').length === 1;
};

let indexOfNotUndefined = (a, start = 0) => {
  let n = a.length;
  for (let i = start; i < n; i++) {
    if (a[i] !== undefined) {
      return i;
    }
  }
  return -1;
};

class Tokenizer {
  constructor(options = {}) {
    options = Object.assign({}, {
      ignore: null,
      patterns: {},
      tokens: [],
    }, options);

    // Validate options
    assert(options.ignore === null || isNonCaptureRegex(options.ignore));
    assert(options.patterns instanceof Object);
    for (let pattern of Object.keys(options.patterns)) {
      assert(isNonCaptureRegex(options.patterns[pattern]));
    }
    assert(options.tokens instanceof Array);
    options.tokens.forEach(tokenName => assert(typeof tokenName === 'string'));

    // Build regular expression
    this._tokens = options.tokens;
    this._hasIgnore = options.ignore ? 1 : 0;
    this._regex = new RegExp('^(?:' + [
      this._hasIgnore ? `(${options.ignore})` : null,
      ...this._tokens.map(tokenName => {
        return `(${options.patterns[tokenName] || escapeRegex(tokenName)})`;
      }),
    ].filter(e => e !== null).join('|') + ')');
  }

  next(source, offset = 0) {
    let m, i;
    do {
      m = this._regex.exec(source.slice(offset));
      if (m === null) {
        // If not at end of input throw an error
        if (source.slice(offset) !== '') {
          throw new SyntaxError(`unexpected EOF for '${source}' at '${source.slice(offset)}'`,
            {start: offset, end: source.length});
        }
        return null;
      }
      i = indexOfNotUndefined(m, 1);
      offset += m[0].length;
    } while (this._hasIgnore && i === 1);
    return {
      kind:   this._tokens[i - 1 - this._hasIgnore],
      value:  m[i],
      start:  offset - m[0].length,
      end:    offset,
    };
  }

  tokenize(source, offset = 0) {
    let token = {end: offset};
    let tokens = [];
    while (token = this.next(source, token.end)) {
      tokens.push(token);
    }
    return tokens;
  }
}

// Export Tokenizer
module.exports = Tokenizer;

File: interpreter/prattparser/tokenizer.go
package prattparser

import (
	"fmt"
	"regexp"
	"strings"
)

// A Token as returend by Tokenizer
type Token struct {
	Kind  string
	Value string
	Start int
	End   int
}

// IsEmpty returns true, if the token is empty. Often used to communicate
// no token is available, such as end of input.
func (t *Token) IsEmpty() bool {
	return t.Kind == "" && t.Value == "" && t.Start == 0 && t.End == 0
}

// A Tokenizer holds rules to tokenize a source string.
type Tokenizer struct {
	kinds     []string
	hasIgnore bool
	regex     *regexp.Regexp
}

// NewTokenizer creates a tokenizer from a language definition
func NewTokenizer(ignore string, tokens []string, patterns map[string]string) *Tokenizer {
	// Simple input validation
	mustBeNonCaptureRegex(ignore)
	for kind, pattern := range patterns {
		if !StringsContains(kind, tokens) {
			panic(fmt.Sprintf("kind '%s' listed in patterns, but not in tokens", kind))
		}
		mustBeNonCaptureRegex(pattern)
	}

	// Consturct options for regular expression
	var options []string
	if ignore != "" {
		options = append(options, fmt.Sprintf(`(%s)`, ignore))
	}
	for _, kind := range tokens {
		pattern, ok := patterns[kind]
		if !ok {
			pattern = regexp.QuoteMeta(kind)
		}
		options = append(options, fmt.Sprintf(`(%s)`, pattern))
	}

	// Construct tokenizer
	return &Tokenizer{
		kinds:     tokens,
		hasIgnore: ignore != "",
		regex:     regexp.MustCompile(fmt.Sprintf(`^(?:%s)`, strings.Join(options, `|`))),
	}
}

// Next returns the next token from source string as given offset
func (t *Tokenizer) Next(source string, offset int) (Token, error) {
	for {
		m := t.regex.FindStringSubmatch(source[offset:])
		if m == nil {
			if source[offset:] != "" {
				return Token{}, SyntaxError{
					Source: source,
					Start:  offset,
					End:    len(source),
				}
			}
			return Token{}, nil
		}
		var i int
		var text string
		for i, text = range m[1:] {
			if text != "" {
				break
			}
		}
		offset += len(text)
		if t.hasIgnore {
			i--
			if i == -1 {
				continue
			}
		}
		return Token{
			Kind:  t.kinds[i],
			Value: text,
			Start: offset - len(text),
			End:   offset,
		}, nil
	}
}

// Tokenize returns a list of tokens from a source string
func (t *Tokenizer) Tokenize(source string, offset int) ([]Token, error) {
	var tokens []Token
	for {
		token, err := t.Next(source, offset)
		if err != nil {
			return nil, err
		}
		if token.IsEmpty() {
			return tokens, nil
		}
		tokens = append(tokens, token)
		offset = token.End
	}
}
